kernel/src/engine/arrow_footer_skipping.rs

```diff
@@ -0,0 +1,359 @@
+        use Expression::*;
+        match expression {
+            VariadicOperation { op, exprs } => self.apply_variadic(op, exprs, inverted),
+            BinaryOperation { op, left, right } => self.apply_binary(op, left, right, inverted),
+            UnaryOperation { op, expr } => self.apply_unary(op, expr, inverted),
+            Literal(value) => Self::apply_scalar(value, inverted),
+            Column(col) => self.apply_column(col, inverted),
+            // We don't support skipping over complex types
+            Struct(_) => None,
+        }
+    }
+
+    fn apply_variadic(
+        &self,
+        op: &VariadicOperator,
+        exprs: &[Expression],
+        inverted: bool,
+    ) -> Option<bool> {
+        let exprs: Vec<_> = exprs
+            .iter()
+            .map(|expr| self.apply_expr(expr, inverted))
+            .collect();
+
+        // With AND (OR), any FALSE (TRUE) input forces FALSE (TRUE) output.  If there was no
+        // dominating input, then any NULL input forces NULL output.  Otherwise, return the
+        // non-dominant value. Inverting the operation also inverts the dominant value.
+        let dominator = match op {
+            VariadicOperator::And => inverted,
+            VariadicOperator::Or => !inverted,
+        };
+        if exprs.iter().any(|v| v.is_some_and(|v| v == dominator)) {
+            Some(dominator)
+        } else if exprs.iter().any(|e| e.is_none()) {
+            None
+        } else {
+            Some(!dominator)
+        }
+    }
+
+    fn apply_binary(
+        &self,
+        op: &BinaryOperator,
+        left: &Expression,
+        right: &Expression,
+        inverted: bool,
+    ) -> Option<bool> {
+        use BinaryOperator::*;
+        use Expression::{Column, Literal};
+
+        // NOTE: We rely on the literal values to provide logical type hints. That means we cannot
+        // perform column-column comparisons, because we cannot infer the logical type to use.
+        let (op, col, val) = match (left, right) {
+            (Column(col), Literal(val)) => (*op, col, val),
+            (Literal(val), Column(col)) => (op.commute()?, col, val),
+            (Literal(a), Literal(b)) => return Self::apply_binary_scalars(op, a, b, inverted),
+            _ => None?, // unsupported combination of operands
+        };
+        let col = col_name_to_path(col);
+        let skipping_eq = |inverted| -> Option<bool> {
+            let below_lo = self.partial_cmp_min_stat(&col, val, Ordering::Less, inverted)?;
+            let above_hi = self.partial_cmp_max_stat(&col, val, Ordering::Greater, inverted)?;
+            let out_of_bounds = below_lo || above_hi;
+            Some(out_of_bounds == inverted)
+        };
+        match op {
+            Equal => skipping_eq(inverted),
+            NotEqual => skipping_eq(!inverted),
+            LessThan => self.partial_cmp_min_stat(&col, val, Ordering::Less, inverted),
```

Comment:
if we re-wrote all the `partial_cmp_[min/max]_stat` calls like:
```rust
 LessThan => partial_cmp_scalars(
                &self.get_min_stat_value(&col, &val.data_type())?,
                val,
                Ordering::Less,
                inverted,
            ),
```

We could remove those functions. It's a little more code at the call site though.

Reply:
Yeah, I was trying to reduce redundancy as much as possible. There's too much anyway that can't be removed.